AppActionVisualizer/app/src/main/java/com/example/appactionvisualizer/ui/adapter/AppRecyclerViewAdapter.java

```diff
@@ -0,0 +1,111 @@
+import android.graphics.drawable.Drawable;
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.ViewGroup;
+import android.widget.ImageView;
+import android.widget.RelativeLayout;
+import android.widget.TextView;
+
+import com.example.appactionvisualizer.R;
+import com.example.appactionvisualizer.constants.Constant;
+import com.example.appactionvisualizer.databean.ActionType;
+import com.example.appactionvisualizer.databean.AppAction;
+import com.example.appactionvisualizer.ui.activity.ActionActivity;
+import com.example.appactionvisualizer.utils.Utils;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+
+/**
+ * Adapter of AppFragment Recyclerview
+ */
+public class AppRecyclerViewAdapter extends RecyclerView.Adapter<AppRecyclerViewAdapter.ViewHolder> {
+
+  private ArrayList<AppAction> appActionArrayList;
+  private Context context;
+
+  public AppRecyclerViewAdapter(ActionType type, Context context) {
+    this.context = context;
+    if(type != null)
+      appActionArrayList = AppAction.appActionList.get(type);
+    else
+      appActionArrayList = AppAction.allAppActions;
+  }
+
+  @Override
+  public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
+    View view = LayoutInflater.from(context)
+        .inflate(R.layout.apptype_rv_item, parent, false);
+    return new ViewHolder(view);
+  }
+
+  @Override
+  public void onBindViewHolder(final ViewHolder holder, final int position) {
+    final AppAction appAction = appActionArrayList.get(position);
+    Drawable icon;
+    try {
+      icon = context.getPackageManager().getApplicationIcon(appAction.getAppPackage());
+      holder.ivAppIcon.setImageDrawable(icon);
+    } catch (PackageManager.NameNotFoundException e) {
+      Utils.showMsg("icon not found", context);
+      e.printStackTrace();
+    }
+    holder.tvAppName.setText(appAction.getAppName());
+    //Use hash set to avoid duplicate tags
+    HashSet<ActionType> uniqueSet = new HashSet<>(5);
+    for (int i = 0; i < appAction.getActions().size(); i++) {
+      uniqueSet.add(appAction.getActions().get(i).getActionType());
+    }
+    int idx = 0;
+    for (ActionType actionType: uniqueSet) {
+      holder.tvAppTags[idx].setVisibility(View.VISIBLE);
+      holder.tvAppTags[idx++].setText(actionType.toString());
+    }
+    holder.rlApp.setOnClickListener(new View.OnClickListener() {
+      @Override
+      public void onClick(View view) {
+        Intent intent = new Intent(context, ActionActivity.class);
+        intent.putExtra(Constant.APP_NAME, appAction);
+        context.startActivity(intent);
+      }
+    });
+  }
+
+  @Override
+  public int getItemCount() {
+    return appActionArrayList.size();
+  }
+
+  public class ViewHolder extends RecyclerView.ViewHolder {
+    public final View mView;
+    public final RelativeLayout rlApp;
+    public final ImageView ivAppIcon;
+    public final TextView tvAppName;
+    public final TextView tvAppTags[] = new TextView[5];
```

Comment:
in google, we usually omit the class type, so instead of ivAppIcon, we will use just "appIcon", you can also prefix all the local variables with the letter "m", such as mAppIcon. You can also omit using "m", up to you, but try to keep naming in this project consistent.

Reply:
Got it